### ISSUE TITLE
Feature Idea: Optimizer/Iterative Interface for pipeline.map

### ISSUE THREAD

**Issue by LennartGevers** [pipefunc/pipefunc#865]
Hi @basnijholt, 

In #513 you've outlined the feature idea to "Have an optimizer interface for calling pipeline.map iteratively." I've encountered several scenarios where such a capability (or iterative behaviour in general) would be incredibly useful, but currently, implementing iterative processes within `pipefunc` pipelines can be cumbersome or require breaking out of the pipefunc paradigm. 

Specifically, I've had occasions where this feature could have significantly streamlined implementations, such as:

- Genetic Algorithm: Iteratively evolving a population, where each generation's fitness evaluation could be a pipeline.map operation, and the selection/crossover/mutation steps feed back into the next iteration.
- Runge-Kutta Methods: Solving ordinary differential equations where each step of the method (e.g., calculating k_1,k_2,k_3,k_4 for RK4) depends on the previous state and feeds forward to compute the next state. 
 
This involves a loop with internal dependencies. The current pipeline.map is excellent for single-pass, N-dimensional sweeps, but it doesn't inherently support the feedback loops necessary for iterative algorithms or optimization routines where the output of one iteration becomes the input for the next, with a defined stopping criterion.

I propose an interface that allows for iterative execution of a sub-pipeline, feeding back parameters from one iteration to the next iterating over a list `xs` (without jeopardizing the concept of a DAG-Pipeline). 

We could introduce a special Pipefunc subclass (e.g., ScanFunc). The user interface would be somewhat simmilar to [`jax.lax.scan`'s](https://docs.jax.dev/en/latest/_autosummary/jax.lax.scan.html#jax.lax.scan) interface. The user would need to specify

1. A `(Nested)PipeFunc` that takes arbitrary input arguments as well as the value of  `xs` at the current index
3. a list `xs` to iterate over (could be provided as `pipeline.run` or `pipeline.map` inputs or by preceding `pipefuncs`, Could be 1-D or in the case of `mapspec` ND but with a certain iteration axis)

1. A Carry `dict` which is merged with the `kwargs` of the previous iteration in the next iteration.
2. (optionally) an intermediate iteration result which is concatenated along with the results of the other iterations's results. (Useful for instance if the iteration produces a time-series like in Runge Kutta Integration, but could be neglected in optimization tasks).

As the number of iterations is given before the execution of the iterative Pipeline Node, we could technically 'unroll' the Node into the Iteration Steps as connected `pipefunc` instances, so that we end up with a linear chain of nodes and a overall `DAG` instead of a `human centipede` like structure in the graph (which we do not want). 
(Although we probably should use a different implementation than that, but I just wanted to clarify that this feature would not break out of the `DAG` paradigm).

```python

@PipeFunc.scan("timeseries", xs="time_steps")
def scan_timeseries(x: float, y: float, t: float, delta_t: float) -> tuple[float, float]

   y_next, new_delta_t = some_rk4_implementation(x, y, delta_t)

   carry = {"y": y,  "delta_t": new_delta_t, "t": t}

   return carry, (y, t)

isinstance(scan_timeseries, ScanFunc) #true
```

> Instead of calling `some_rk4_implementation` one could also create a `ScanFunc` with `NestedPipeFunc.scan` and use the pipeline paradigm for the rk4 implementation.

Currently, to achieve iterative behavior, one typically has to:

- Run a pipefunc pipeline to get an output.
- Take that output, perform some external looping logic in Python.
- Feed the modified output back into another pipefunc call (or the same pipeline instance).

This breaks the seamless composition and visualization benefits that pipefunc provides, making the overall workflow less declarative and harder to manage, especially for parallel execution and N-dimensional sweeps within the loop. Furthermore, any pre- or post-processing tasks of the results would need to be implemented in a separate pipeline instance, leading to a fragmented and potentially awkward hierarchy of nested pipelines.

**Additional context**

This feature would significantly enhance pipefunc's utility for a broader range of scientific computing and machine learning tasks that inherently involve iterative optimization, simulation, or refinement processes. It aligns well with the library's goal of simplifying complex workflows and focusing on the science rather than execution management.

**Comment by basnijholt** [comment]
Sorry for my slow response. I think your idea sounds great. However, I think something is missing from your example because it doesn't show what maps the `xs` parameter to `x`.

**Comment by basnijholt** [comment]
Just to test, I started on this using Claude Code and we got quite far in just a few hours: https://github.com/pipefunc/pipefunc/pull/867.

Still draft and will need a lot of improvements but looks promising!

**Comment by basnijholt** [comment]
Sorry for my slow response. I think your idea sounds great.

Just to test, I started on this using Claude Code and we got quite far in just a few hours: https://github.com/pipefunc/pipefunc/pull/867.

Still draft and will need a lot of improvements but looks promising! 

I've implemented iterative execution support in two ways:

## 1. ScanFunc (JAX-style)

Following your proposed interface:

```python
@PipeFunc.scan(output_name="timeseries", xs="time_steps")
def scan_timeseries(x: float, y: float, delta_t: float, t: float = 0.0) -> tuple[dict[str, Any], float]:
    y_next, new_delta_t = some_rk4_implementation(x, y, delta_t)
    carry = {"y": y_next, "delta_t": new_delta_t, "t": t + delta_t}
    return carry, y_next
```

## 2. scan_iter (Generator-based)

A more Pythonic alternative:

```python
@PipeFunc.scan_iter(output_name="timeseries")
def scan_timeseries(time_steps: list[float], y: float = 1.0, delta_t: float = 0.01):
    t = 0.0
    for x in time_steps:
        y, delta_t = some_rk4_implementation(x, y, delta_t)
        t += delta_t
        yield {"y": y, "t": t}
```

## Your Use Cases

**Genetic Algorithm:**
```python
@PipeFunc.scan_iter(output_name="evolution")
def genetic_algorithm(generations: int, population_size: int = 100):
    population = initialize_population(population_size)
    for gen in range(generations):
        fitness = evaluate_fitness(population)
        population = evolve(population, fitness)
        yield {"generation": gen, "best_fitness": np.max(fitness)}
```

**RK4:**
```python
@PipeFunc.scan_iter(output_name="trajectory")
def rk4(time_points: np.ndarray, y0: float, f: Callable):
    y = y0
    dt = time_points[1] - time_points[0]
    for t in time_points:
        k1 = dt * f(t, y)
        k2 = dt * f(t + 0.5*dt, y + 0.5*k1)
        k3 = dt * f(t + 0.5*dt, y + 0.5*k2)
        k4 = dt * f(t + dt, y + k3)
        y = y + (k1 + 2*k2 + 2*k3 + k4) / 6
        yield {"t": t, "y": y}
```

Both support:
- Carry state between iterations
- Optional intermediate results collection
- Full pipeline.map integration
- Nested pipeline execution
- Resource specifications


I actually think the second version is nicer because it also supports breaking early, for example.

**Comment by basnijholt** [comment]
OK sorry for all the noise 😅

Now that I spent a couple of hours implementing this I took a step back to think whether this enables anything new.

Here's the Runge-Kutta example implemented both ways:

```python
# With scan_iter
@PipeFunc.scan_iter(output_name="trajectory")
def rk4_scan(time_steps: list[float], y0: float, dt: float):
    y = y0
    for t in time_steps:
        k1 = dt * dydt(t, y)
        k2 = dt * dydt(t + dt/2, y + k1/2)
        k3 = dt * dydt(t + dt/2, y + k2/2)
        k4 = dt * dydt(t + dt, y + k3)
        y = y + (k1 + 2*k2 + 2*k3 + k4) / 6
        yield {"t": t, "y": y}

# Without scan - just regular @pipefunc
@pipefunc(output_name="trajectory")
def rk4_regular(time_steps: list[float], y0: float, dt: float) -> list[dict]:
    results = []
    y = y0
    for t in time_steps:
        k1 = dt * dydt(t, y)
        k2 = dt * dydt(t + dt/2, y + k1/2)
        k3 = dt * dydt(t + dt/2, y + k2/2)
        k4 = dt * dydt(t + dt, y + k3)
        y = y + (k1 + 2*k2 + 2*k3 + k4) / 6
        results.append({"t": t, "y": y})
    return results
```

Both work identically in pipelines, with `pipeline.map()`, and maintain the DAG structure.

But ... maybe I am missing the whole point?

**Comment by LennartGevers** [comment]
First of all, thanks for working on the feature @basnijholt!

> Sorry for my slow response.

no problem. I initially intended to work on this feature myself, but could not manage to find the spare time.

> Both work identically in pipelines, with pipeline.map(), and maintain the DAG structure.
> 
> But ... maybe I am missing the whole point?

So the `Pipefunc.iter_scan` feature would be about 90% redundant to the simple for-loop implementation . The main benefit would be that progress of very long loops could be tracked and integrated into the progress tracking of `mapspec` operations.

**The main idea** lies within a `NestedPipeFunc.iter_scan` implementation. If we take the Genetic Algorithm example for instance,

```python
@PipeFunc.scan_iter(output_name="evolution")
def genetic_algorithm(generations: int, population_size: int = 100):
    population = initialize_population(population_size)
    for gen in range(generations):
        fitness = evaluate_fitness(population)
        population = evolve(population, fitness)
        yield {"generation": gen, "best_fitness": np.max(fitness)}
```

we could also express this workflow in a way that resembles more of the `pipefunc`-style, possibly enabling features, like `mapspec`, `variants`

```python

@pipefunc("initial_population")
def initialize_population(population_size: int, genes: dict[str, tuple[float, float]]) -> ...:
   ...

@pipefunc("fitness", mapspec="..., ... -> ...")
def fitness(population: ..., ...) -> ...:
   # This might also be an entire workflow like a simulation consisting of serveral pipefuncs
   # or a NestedPipeFunc
   ...

@pipefunc("best_solution")
def pick_best_solution(population: ..., fitness: ...) -> ...:
   ...

@pipefunc("parents", variant={"parent_selection": "roulette_wheel"})
def parent_selection(carry: ..., fitness: ..., ...) -> ...:
   ...

@pipefunc("offsprings", variant{"crossover": "single_point"})
def single_point_crossover(parents: ..., ...) -> ...:
   ...

@pipefunc("mutated_offsprings")
def mutation(offsprings: ..., generation: int, ...) -> ...: 
   # we take generation as a input here for a deterministic random number seed
   ...

@pipefunc("updated_population")
def updated_population(population: ..., mutated_offsprings: ...) -> ...:
   ...

@pipefunc("generations")
def generations(num_generations: int) -> list[int]:
   return list(range(num_generations))

genetic_algorithm_func = NestedPipeFunc.scan_iter(
   [
      fitness,
      parent_selection,
      single_point_crossover,
      mutation,
      updated_population,
      pick_best_solution,
      generations,
   ],
   xs="generations",
   initial_carry="initial_population",
   carry_updater="updated_population",
   iteration_output="best_solution",
   output_name="generations",
)
```

With the first implementation, we almost entirely leave the `pipefunc`-'zen' and resort to basic python, leaving features like composable workflows, mapspec, variants, caching, parallelization and so on behind. With the `NestedPipeFunc.scan_iter` implementation, we could build upon most of these features (tho this might require some work on variants within NestedPipeFuncs).

> I actually think the second (yield) version is nicer because it also supports breaking early, for example.

I think this might result in the same problems as #718, if we also want to support `mapspec`.
